File: src/main/java/Solution_383_RansomNote.java
import java.util.HashMap;

public class Solution_383_RansomNote {
    public static void main(String[] args) {
        var ransomNote = "aa";
        var magazine = "ab";
        canConstruct(ransomNote, magazine);
    }

    private static boolean canConstruct(String ransomNote, String magazine) {
        var lettersCount = new HashMap<Character, Integer>();

        for (var letter : magazine.toCharArray()) {
            lettersCount.merge(letter, 1, Integer::sum);
//            lettersCount.merge(letter, 1, (oldValue, newValue) -> oldValue + newValue);
//            lettersCount.compute(letter, (key, value) -> value == null ? 1 : value + 1);
//            lettersCount.put(letter, lettersCount.getOrDefault(letter, 0) + 1);
        }

        for (var letter : ransomNote.toCharArray()) {
            var count = lettersCount.get(letter);
            if (count == null || count == 0) return false;
            lettersCount.compute(letter, (key, value) -> value - 1);
        }

        return true;
    }
}

File: src/main/kotlin/Solution_383_RansomNote.kt
/**
 * 383 Ransom Note
 *
 * Easy
 *
 * Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the letters from magazine and false otherwise.
 *
 * Each letter in magazine can only be used once in ransomNote.
 *
 *
 *
 * Example 1:
 *
 * Input: ransomNote = "a", magazine = "b"
 * Output: false
 * Example 2:
 *
 * Input: ransomNote = "aa", magazine = "ab"
 * Output: false
 * Example 3:
 *
 * Input: ransomNote = "aa", magazine = "aab"
 * Output: true
 *
 *
 * Constraints:
 *
 * 1 <= ransomNote.length, magazine.length <= 105
 * ransomNote and magazine consist of lowercase English letters.
 */
fun main() {
    val ransomNote = "aa"
    val magazine = "ab"
    canConstruct(ransomNote, magazine)
}

/**
 * можно ли из строки magazine получить строку ransomNote
 */
fun canConstruct(ransomNote: String, magazine: String): Boolean {
    // считаем количество вхождений каждого символа в magazine
    val lettersCount = mutableMapOf<Char, Int>()

    for (letter in magazine) {
//        lettersCount.compute(letter) { key, value -> (value ?: 0) + 1 }
//        lettersCount.merge(letter, 1) { oldValue, newValue -> oldValue + newValue }
//        lettersCount.merge(letter, 1, Int::plus)

        lettersCount[letter] = lettersCount.getOrDefault(letter, 0) + 1 // так быстрее всего
    }

    // проверяем каждый символ из letter, что он есть в magazine и уменьшаем счетчик
    // если символа нет или его количество равно 0, возвращаем false
    for (letter in ransomNote) {
        var count = lettersCount[letter] ?: return false
        if (count == 0) return false
        lettersCount[letter] = --count
    }

    return true
}